File: array/42. Trapping Rain Water.py
from typing import List

# Array
#! Time : O(n) ; Space : O(n)
class Solution:
    def trap(self, height: List[int]) -> int:
        result, rMax, lMax = 0, [0] * len(height), [0] * len(height)
        temp = height[0]
        for i in range(0, len(height)):
            if temp < height[i]:
                temp = lMax[i] = height[i]
            else:
                lMax[i] = temp
        temp = height[-1]
        for i in range(0, len(height)):
            if temp < height[-1 - i]:
                rMax[-1 - i] = temp = height[-1 - i]
            else:
                rMax[-1 - i] = temp
        for i in range(0, len(height)):
            result += max(0, min(lMax[i], rMax[i]) - height[i])
        return result
        

File: two pointer/42. Trapping Rain Water.ts
// Two pointer
// Time : O(n) ; Space : O(1) 

function trap(height: number[]): number {
    let n = height.length, result = 0;
    let l = 0, r = n - 1;
    let lMax = height[0], rMax = height[n - 1];
    while (l < r) {
        if (height[l] <= height[r]) {
            result += Math.max(0, lMax - height[l]);
            lMax = Math.max(lMax, height[l]);
            l++;
        } else {
            result += Math.max(0, rMax - height[r]);
            rMax = Math.max(rMax, height[r]);
            r--;
        }
    }
    return result;
};